File: art-factory/fortune-v2-20260716/factory-qa.mjs
import { createRequire } from 'node:module';
import { readFile, writeFile, mkdir, readdir } from 'node:fs/promises';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const require = createRequire(import.meta.url);
const sharp = require('sharp');
const root = path.dirname(fileURLToPath(import.meta.url));
const manifest = JSON.parse(await readFile(path.join(root, 'art-manifest.json'), 'utf8'));
const source = path.resolve(root, manifest.geometryReference);
const outputs = path.join(root, manifest.outputRoot);
const previews = path.join(root, 'previews');
const canvas = manifest.canvas[0];
await mkdir(previews, { recursive: true });

const ordered = [...manifest.layers['obsidian-gold']].sort((a, b) => a.z - b.z);

async function exists(file) {
  try { await sharp(file).metadata(); return true; } catch { return false; }
}

async function assembled({ face, skinDir, extra = [], background = '#f4eee4' } = {}) {
  const layers = [];
  for (const layer of ordered) {
    let file = path.join(source, `${layer.id}.png`);
    if (layer.id === 'pet-face' && face) file = face;
    if (skinDir && manifest.layers['ivory-vermilion'].includes(layer.id)) {
      const candidate = path.join(skinDir, `${layer.id}.png`);
      if (await exists(candidate)) file = candidate;
    }
    layers.push({ input: file, left: 0, top: 0, z: layer.z });
  }
  layers.push(...extra.map((item) => typeof item === 'string'
    ? { input: item, left: 0, top: 0, z: 100 }
    : { input: item.input, left: 0, top: 0, z: item.z ?? 100 }));
  layers.sort((a, b) => a.z - b.z);
  return sharp({
    create: { width: canvas, height: canvas, channels: 4, background },
  }).composite(layers.map(({ z, ...layer }) => layer)).png().toBuffer();
}

async function label(text, width, height) {
  const safe = text.replaceAll('&', '&amp;').replaceAll('<', '&lt;');
  return Buffer.from(`<svg width="${width}" height="${height}" xmlns="http://www.w3.org/2000/svg"><rect width="100%" height="100%" rx="18" fill="#171c2e"/><text x="24" y="44" fill="#f4eee4" font-family="system-ui,sans-serif" font-size="24" font-weight="700">${safe}</text></svg>`);
}

async function contactSheet(cells, outFile, columns = 2) {
  const tile = 512, header = 64;
  const rows = Math.ceil(cells.length / columns);
  const composites = [];
  for (let i = 0; i < cells.length; i++) {
    const x = (i % columns) * tile, y = Math.floor(i / columns) * (tile + header);
    const thumb = await sharp(cells[i].image).resize(tile, tile, { fit: 'contain' }).png().toBuffer();
    composites.push({ input: await label(cells[i].label, tile, header), left: x, top: y });
    composites.push({ input: thumb, left: x, top: y + header });
  }
  await sharp({
    create: { width: columns * tile, height: rows * (tile + header), channels: 4, background: '#0f1320' },
  }).composite(composites).png().toFile(outFile);
}

async function validate(file) {
  const image = sharp(file).ensureAlpha();
  const { data, info } = await image.raw().toBuffer({ resolveWithObject: true });
  let nonzero = 0, fringe = 0, minX = info.width, minY = info.height, maxX = -1, maxY = -1;
  const alphaAt = (x, y) => data[(y * info.width + x) * 4 + 3];
  for (let y = 0; y < info.height; y++) for (let x = 0; x < info.width; x++) {
    const i = (y * info.width + x) * 4, a = data[i + 3];
    if (!a) continue;
    nonzero++;
    minX = Math.min(minX, x); minY = Math.min(minY, y); maxX = Math.max(maxX, x); maxY = Math.max(maxY, y);
    if (data[i + 1] > Math.max(data[i], data[i + 2]) + 32 && data[i + 1] > 92) fringe++;
  }
  return {
    file: path.relative(root, file),
    size: [info.width, info.height],
    alpha: info.channels === 4,
    transparentCorners: [[0,0],[info.width-1,0],[0,info.height-1],[info.width-1,info.height-1]].every(([x,y]) => alphaAt(x,y) === 0),
    coverage: Number((nonzero / (info.width * info.height)).toFixed(6)),
    bbox: nonzero ? [minX, minY, maxX + 1, maxY + 1] : null,
    chromaFringePixels: fringe,
  };
}

async function exactAlphaMatch(file, reference) {
  const [candidate, approved] = await Promise.all([
    sharp(file).ensureAlpha().extractChannel(3).raw().toBuffer(),
    sharp(reference).ensureAlpha().extractChannel(3).raw().toBuffer(),
  ]);
  return candidate.equals(approved);
}

async function isolatedPreview(file, { height } = {}) {
  let layer;
  if (height) layer = await sharp(file).resize({ height, fit: 'contain' }).png().toBuffer();
  else layer = await sharp(file).png().toBuffer();
  const metadata = await sharp(layer).metadata();
  return sharp({
    create: { width: canvas, height: canvas, channels: 4, background: '#f4eee4' },
  }).composite([{
    input: layer,
    left: Math.round((canvas - metadata.width) / 2),
    top: Math.round((canvas - metadata.height) / 2),
  }]).png().toBuffer();
}

const stateFaces = {
  hungry: path.join(outputs, 'states/pet-face-hungry.png'),
  thriving: path.join(source, 'pet-face.png'),
  full: path.join(outputs, 'states/pet-face-full.png'),
  overfed: path.join(outputs, 'states/pet-face-overfed.png'),
};
if ((await Promise.all(Object.values(stateFaces).map(exists))).every(Boolean)) {
  const cells = [];
  for (const [id, face] of Object.entries(stateFaces)) {
    const image = await assembled({ face });
    await sharp(image).toFile(path.join(previews, `obsidian-gold-${id}.png`));
    cells.push({ label: id, image });
  }
  await contactSheet(cells, path.join(previews, 'batch-a-states-contact-sheet.png'));
}

const skinDir = path.join(outputs, 'skins/ivory-vermilion');
const skinFiles = manifest.layers['ivory-vermilion'].map((id) => path.join(skinDir, `${id}.png`));
if ((await Promise.all(skinFiles.map(exists))).every(Boolean)) {
  const ivory = await assembled({ skinDir });
  await sharp(ivory).toFile(path.join(previews, 'ivory-vermilion-thriving.png'));
  const obsidian = await assembled();
  await contactSheet([
    { label: 'obsidian-gold', image: obsidian },
    { label: 'ivory-vermilion', image: ivory },
  ], path.join(previews, 'batch-b-skins-comparison.png'), 2);
  const cells = [];
  for (let i = 0; i < skinFiles.length; i++) {
    const image = await sharp({ create: { width: canvas, height: canvas, channels: 4, background: '#f4eee4' } })
      .composite([{ input: skinFiles[i], left: 0, top: 0 }]).png().toBuffer();
    cells.push({ label: manifest.layers['ivory-vermilion'][i], image });
  }
  await contactSheet(cells, path.join(previews, 'batch-b-ivory-layers-contact-sheet.png'), 4);
}

const props = [
  { id: 'prop-counting-coins', z: 100 },
  { id: 'prop-stash-bag', z: 25 },
  { id: 'prop-ledger-glasses', z: 100 },
];
const propFiles = props.map((p) => path.join(outputs, `props/${p.id}.png`));
if ((await Promise.all(propFiles.map(exists))).every(Boolean)) {
  const cells = [];
  for (let i = 0; i < props.length; i++) {
    const image = await assembled({ extra: [{ input: propFiles[i], z: props[i].z }] });
    await sharp(image).toFile(path.join(previews, `${props[i].id}-assembled.png`));
    cells.push({ label: props[i].id, image });
  }
  await contactSheet(cells, path.join(previews, 'batch-c-props-contact-sheet.png'), 3);
}

const wearables = [
  { id: 'head-luck-crown', slot: 'head', z: 95 },
  { id: 'head-accountant-glasses', slot: 'head', z: 100 },
  { id: 'neck-ceremonial-bow', slot: 'neck', z: 75 },
  { id: 'neck-jade-amulet', slot: 'neck', z: 75 },
  { id: 'back-merchant-scroll', slot: 'back', z: 25 },
  { id: 'back-festival-cloak', slot: 'back', z: 25 },
];
const wearableFiles = wearables.map((item) => path.join(outputs, `wearables/${item.slot}/${item.id}.png`));
if ((await Promise.all(wearableFiles.map(exists))).every(Boolean)) {
  const cells = [];
  for (let i = 0; i < wearables.length; i++) {
    const item = wearables[i];
    const image = await assembled({ extra: [{ input: wearableFiles[i], z: item.z }] });
    await sharp(image).toFile(path.join(previews, `${item.id}-assembled.png`));
    cells.push({ label: `${item.slot}: ${item.id}`, image });
  }
  await contactSheet(cells, path.join(previews, 'batch-d-wearables-contact-sheet.png'), 3);
}

const shadowFile = path.join(outputs, 'shadow/shadow-spirit-calm.png');
const currentShadow = path.resolve(root, '../../public/assets/shadow/shadow_2.png');
if (await exists(shadowFile) && await exists(currentShadow)) {
  const current = await isolatedPreview(currentShadow, { height: 570 });
  const spiritCalm = await isolatedPreview(shadowFile);
  await sharp(spiritCalm).toFile(path.join(previews, 'shadow-spirit-calm.png'));
  await contactSheet([
    { label: 'current shadow identity', image: current },
    { label: 'style check: spirit + calm', image: spiritCalm },
  ], path.join(previews, 'batch-e-shadow-spirit-calm.png'), 2);
}

const candidates = [];
async function walk(dir) {
  for (const ent of await readdir(dir, { withFileTypes: true })) {
    const file = path.join(dir, ent.name);
    if (ent.isDirectory()) await walk(file);
    else if (ent.name.endsWith('.png')) candidates.push(file);
  }
}
await walk(outputs);
const report = await Promise.all(candidates.sort().map(validate));
const approvedFaceGeometry = await validate(path.join(source, 'pet-face.png'));
for (const asset of report) {
  const skinMatch = asset.file.match(/^outputs\/skins\/ivory-vermilion\/(.+)\.png$/);
  const stateMatch = asset.file.match(/^outputs\/states\/(.+)\.png$/);
  if (skinMatch) {
    asset.geometryMode = 'exact-alpha';
    asset.geometryMatch = await exactAlphaMatch(path.join(root, asset.file), path.join(source, `${skinMatch[1]}.png`));
  }
  if (stateMatch) {
    asset.geometryMode = 'exact-master-bbox';
    asset.geometryMatch = JSON.stringify(asset.bbox) === JSON.stringify(approvedFaceGeometry.bbox);
  }
  asset.pass = asset.size[0] === canvas && asset.size[1] === canvas && asset.alpha && asset.transparentCorners
    && asset.coverage > 0 && asset.chromaFringePixels <= 512 && asset.geometryMatch !== false;
}
const qa = {
  generatedAt: new Date().toISOString(),
  assetCount: report.length,
  passed: report.filter((asset) => asset.pass).length,
  failed: report.filter((asset) => !asset.pass).length,
  assets: report,
};
await writeFile(path.join(root, 'qa-report.json'), `${JSON.stringify(qa, null, 2)}\n`);
const md = [
  '# Fortune Cat v2 factory QA',
  '',
  `- Assets: ${qa.assetCount}`,
  `- Passed: ${qa.passed}`,
  `- Failed: ${qa.failed}`,
  '- Contract: 1024×1024 RGBA, transparent corners, non-empty alpha, no material chroma fringe.',
  '- Recolored skin/state layers additionally require pixel-identical approved alpha geometry.',
  '',
  '| asset | bbox | coverage | chroma fringe | geometry | result |',
  '|---|---:|---:|---:|---:|---:|',
  ...report.map((asset) => `| ${asset.file} | ${asset.bbox?.join(', ') ?? 'empty'} | ${asset.coverage} | ${asset.chromaFringePixels} | ${asset.geometryMatch === undefined ? 'n/a' : asset.geometryMatch ? asset.geometryMode : 'FAIL'} | ${asset.pass ? 'PASS' : 'FAIL'} |`),
  '',
];
await writeFile(path.join(root, 'qa-report.md'), `${md.join('\n')}\n`);

const overviewSources = [
  ['A · state faces', 'batch-a-states-contact-sheet.png'],
  ['B · skin variant', 'batch-b-skins-comparison.png'],
  ['C · props', 'batch-c-props-contact-sheet.png'],
  ['D · wearables', 'batch-d-wearables-contact-sheet.png'],
  ['E · shadow style check', 'batch-e-shadow-spirit-calm.png'],
].map(([label, file]) => ({ label, file: path.join(previews, file) }));
if ((await Promise.all(overviewSources.map(({ file }) => exists(file)))).every(Boolean)) {
  await contactSheet(overviewSources.map(({ label, file }) => ({ label, image: file })), path.join(previews, 'factory-overview.png'), 2);
}
console.log(`QA: ${report.length} assets; previews: ${previews}`);

File: art-factory/traveller-v1-20260725/factory-qa.py
#!/usr/bin/env python3
"""Validate Traveller production layers and build current assembly previews."""

from __future__ import annotations

import json
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from PIL import Image, ImageDraw, ImageFont


ROOT = Path(__file__).resolve().parent
MANIFEST = json.loads((ROOT / "art-manifest.json").read_text())
WIDTH, HEIGHT = MANIFEST["canvas"]
OUTPUTS = ROOT / MANIFEST["outputRoot"]
PREVIEWS = ROOT / "previews"
CONTROL = ROOT / MANIFEST["assembledReference"]
PAPER = (244, 238, 228, 255)
INK = (23, 28, 46, 255)
PREVIEWS.mkdir(parents=True, exist_ok=True)


def resolve_layer(layer: dict[str, Any]) -> Path:
    return OUTPUTS / layer["file"]


def validate(file: Path) -> dict[str, Any]:
    original = Image.open(file)
    has_alpha = original.mode in {"RGBA", "LA"} or "transparency" in original.info
    image = original.convert("RGBA")
    pixels = image.load()
    nonzero = 0
    fringe = 0
    min_x, min_y = image.width, image.height
    max_x = max_y = -1

    for y in range(image.height):
        for x in range(image.width):
            red, green, blue, alpha = pixels[x, y]
            if alpha == 0:
                continue
            nonzero += 1
            min_x = min(min_x, x)
            min_y = min(min_y, y)
            max_x = max(max_x, x)
            max_y = max(max_y, y)
            if green > max(red, blue) + 32 and green > 92:
                fringe += 1

    corners = [
        pixels[0, 0][3],
        pixels[image.width - 1, 0][3],
        pixels[0, image.height - 1][3],
        pixels[image.width - 1, image.height - 1][3],
    ]
    result = {
        "file": str(file.relative_to(ROOT)),
        "size": [image.width, image.height],
        "alpha": has_alpha,
        "transparentCorners": all(alpha == 0 for alpha in corners),
        "coverage": round(nonzero / (image.width * image.height), 6),
        "bbox": [min_x, min_y, max_x + 1, max_y + 1] if nonzero else None,
        "chromaFringePixels": fringe,
    }
    result["pass"] = (
        result["size"] == [WIDTH, HEIGHT]
        and result["alpha"]
        and result["transparentCorners"]
        and result["coverage"] > 0
        and result["chromaFringePixels"] <= 1024
    )
    return result


def paper_canvas() -> Image.Image:
    return Image.new("RGBA", (WIDTH, HEIGHT), PAPER)


def composite_layers(layers: list[dict[str, Any]]) -> Image.Image:
    canvas = paper_canvas()
    for layer in layers:
        canvas.alpha_composite(Image.open(layer["file"]).convert("RGBA"), (0, 0))
    return canvas


def font(size: int) -> ImageFont.ImageFont:
    candidates = [
        Path("/System/Library/Fonts/SFNS.ttf"),
        Path("/System/Library/Fonts/Supplemental/Arial Bold.ttf"),
    ]
    for candidate in candidates:
        if candidate.exists():
            return ImageFont.truetype(str(candidate), size)
    return ImageFont.load_default()


def contact_sheet(
    cells: list[tuple[str, Image.Image]],
    output: Path,
    columns: int = 3,
) -> None:
    if not cells:
        return
    tile_width, tile_height, header = 384, 576, 56
    rows = (len(cells) + columns - 1) // columns
    sheet = Image.new("RGBA", (columns * tile_width, rows * (tile_height + header)), INK)
    draw = ImageDraw.Draw(sheet)
    label_font = font(22)

    for index, (label, image) in enumerate(cells):
        x = (index % columns) * tile_width
        y = (index // columns) * (tile_height + header)
        thumb = image.copy()
        thumb.thumbnail((tile_width, tile_height), Image.Resampling.LANCZOS)
        left = x + (tile_width - thumb.width) // 2
        top = y + header + (tile_height - thumb.height) // 2
        sheet.alpha_composite(thumb, (left, top))
        draw.text((x + 16, y + 15), label, fill=PAPER, font=label_font)
    sheet.save(output)


ordered = sorted(MANIFEST["layers"], key=lambda layer: layer["z"])
present: list[dict[str, Any]] = []
for layer in ordered:
    file = resolve_layer(layer)
    if file.exists():
        present.append({**layer, "file": file})

current = composite_layers(present)
current.save(PREVIEWS / "current-assembled.png")

isolated_cells: list[tuple[str, Image.Image]] = []
for layer in present:
    isolated = paper_canvas()
    isolated.alpha_composite(Image.open(layer["file"]).convert("RGBA"), (0, 0))
    isolated_cells.append((f'{layer["id"]} · z{layer["z"]}', isolated))
contact_sheet(isolated_cells, PREVIEWS / "current-layers-contact-sheet.png")

if CONTROL.exists():
    approved = paper_canvas()
    approved.alpha_composite(Image.open(CONTROL).convert("RGBA"), (0, 0))
    contact_sheet(
        [
            ("approved control master", approved),
            (f"current split · {len(present)} layers", current),
        ],
        PREVIEWS / "control-vs-current.png",
        columns=2,
    )

assets = [validate(layer["file"]) for layer in present]
report = {
    "generatedAt": datetime.now(timezone.utc).isoformat(),
    "canvas": MANIFEST["canvas"],
    "expectedLayerCount": len(MANIFEST["layers"]),
    "presentLayerCount": len(present),
    "passed": sum(asset["pass"] for asset in assets),
    "failed": sum(not asset["pass"] for asset in assets),
    "assets": assets,
}
(ROOT / "qa-report.json").write_text(json.dumps(report, ensure_ascii=False, indent=2) + "\n")

lines = [
    "# Traveller v1 factory QA",
    "",
    f"- Canvas: {WIDTH}×{HEIGHT}",
    f'- Present layers: {report["presentLayerCount"]}/{report["expectedLayerCount"]}',
    f'- Passed: {report["passed"]}',
    f'- Failed: {report["failed"]}',
    "- Contract: full master canvas, real alpha, transparent corners, non-empty layer, no material green fringe.",
    "",
    "| asset | bbox | coverage | chroma fringe | result |",
    "|---|---:|---:|---:|---:|",
]
for asset in assets:
    bbox = ", ".join(map(str, asset["bbox"])) if asset["bbox"] else "empty"
    lines.append(
        f'| {asset["file"]} | {bbox} | {asset["coverage"]} | '
        f'{asset["chromaFringePixels"]} | {"PASS" if asset["pass"] else "FAIL"} |'
    )
lines.append("")
(ROOT / "qa-report.md").write_text("\n".join(lines))
print(
    f'Traveller QA: {report["presentLayerCount"]}/{report["expectedLayerCount"]} '
    f'layers; {report["failed"]} failures'
)
